### DISCUSSION TITLE
Subscription - ApolloServer context & Graphql-ws context

### DISCUSSION THREAD

**Discussion by RemyMachado** [apollographql/apollo-server#7081] [Q&A]
I'm currently setting the apollo server context using an async callback:

```typescript
const server = new ApolloServer({
 ...
 context: ({ req }) => {
        ... // returning Promise<context>
        }
}
```

Now I'm duplicating the code to give a context to the `graphql-ws` implementation:
```typescript
const webSocketServerCleanup = useWebSocketServer(
        {
            schema: schemasAndResolvers,
            context: {
                ...
            },
        },
        wsServer
    )
```

I didn't succeed to set it up the same way since the signatures are different (how to retrieve the req obj for WS?). 
ApolloServer context:
`context?: Context | ContextFunction<ContextFunctionParams>;`  

Graphql-ws context:
`context?: GraphQLExecutionContextValue | ((ctx: Context<P, E>, message: SubscribeMessage, args: ExecutionArgs) => Promise<GraphQLExecutionContextValue> | GraphQLExecutionContextValue);`

### What's the right way to have the same context in both servers?

Any help would be greatly appreciated.

**Comment by RemyMachado** [comment]
I found that I can use `ctx.connectionParams.Authorization` to retrieve the bearerToken, and for the dataSources I may call the creatorFn, but I'm not sure it's the best way. Isn't there a way to synchronize them, without creating them twice?

**Comment by glasser** [comment]
You do have to write different code because the data structures are actually different. One represents parsing an HTTP request (based on your favorite web framework such as Request); the other represents parsing the beginning of a Websocket stream.

**Reply by RemyMachado** [reply]
Thank you for your answer, it makes sense.

I'll leave the discussion unanswered, in case I or someone come up with a solid implementation. I might create an adapter to reuse most of the logic and create an example.
